[arrow] Update to 4.0.0

ianmcook opened the pull request:
**Describe the pull request**

- #### What does your PR fix?  
  Fixes #17635 in vcpkg issues
  Fixes [ARROW-11581](https://issues.apache.org/jira/browse/ARROW-11581) in the Apache Arrow Jira

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  Not supported: arm-uwp, x86-windows, arm64-windows
  Yes, CI baseline is updated

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  Yes

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

-----

@GPSnoopy please take a look.

~~There are no patches yet. The patch-related lines in `portfile.cmake` are commented out.~~

This installs successfully for me on macOS and Windows. ~~On Windows I'm hitting a problem at the `vcpkg_install_cmake()` step that I'm still trying to debug.~~

I invited you as a collaborator on my fork, so you should be able to push commits to this PR after you accept the invite.

Thanks!

ianmcook committed:
Update arrow to 4.0.0

ianmcook committed:
Format

ianmcook committed:
Try fix thrift

ianmcook committed:
Update versions/ files

ianmcook committed:
Do not set ZSTD_ROOT

ianmcook committed:
Remove double quotes causing Windows problems

ianmcook commented:
Some notes on what I changed so far:
- I switched to using a `vcpkg.json` manifest
- I enabled several new features: `dataset`, `flight`, `orc`, `s3`, and the two optional allocators `jemalloc` and `mimalloc`
  - `mimalloc` is on by default on Windows and `jemalloc` is on by default on other OSs
  - The other on-by-default features are the same as in the 3.0.0 port, except for the `dataset` module which is used increasingly
  - I declared feature-specific dependencies within those features only so they will not be installed when the features are off
- I alphabetized all the CMake options
- I removed all the individual `_USE_SHARED` in favor of `ARROW_DEPENDENCY_USE_SHARED` which controls them all, leaving only Thrift controlled independently
- I removed `double-conversion`, `flatbuffers`, and `uriparser` because they are vendored in Arrow now
- I removed `boost-algorithm` because I believe it is not needed any longer

ianmcook commented on ports/arrow/portfile.cmake:
```diff
@@ -1,64 +1,98 @@
+    set(MALLOC_OPTIONS ${MALLOC_OPTIONS} -DARROW_MIMALLOC=OFF)
+endif()
+
+set(LIB_DIR_OPTIONS
+    -DBOOST_ROOT=${CURRENT_INSTALLED_DIR}
+    -DBOOST_INCLUDEDIR=${CURRENT_INSTALLED_DIR}/include/boost
+    -DBOOST_LIBRARYDIR=${CURRENT_INSTALLED_DIR}/lib
+    -DOPENSSL_INCLUDE_DIR=${CURRENT_INSTALLED_DIR}/include
+    -DOPENSSL_LIBRARIES=${CURRENT_INSTALLED_DIR}/lib
+    -DOPENSSL_ROOT_DIR=${CURRENT_INSTALLED_DIR}
+    -DThrift_ROOT=${CURRENT_INSTALLED_DIR}/lib
+    -DZSTD_INCLUDE_DIR=${CURRENT_INSTALLED_DIR}/include
 )
+if(VCPKG_HOST_IS_WINDOWS)
+    set(LIB_DIR_OPTIONS
+        ${LIB_DIR_OPTIONS}
+        -DLZ4_MSVC_LIB_PREFIX=
+        -DLZ4_MSVC_STATIC_LIB_SUFFIX=
+        -DZSTD_MSVC_LIB_PREFIX=
+    )
+endif()
```
I hoped this could reduce the need for patches, but I'm not convinced it does

GPSnoopy commented:
Wow @ianmcook. Impressive work.

It looks like some of the CI/CD errors are unrelated to your commit: dbow2, wxwidgets, embree3, mlpack, zxing-cpp. Looks like you might have branched out from a broken master. Worth merging master into your branch, it looks green for the moment (and some of these packages got updated quite recently).

GPSnoopy committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg

GPSnoopy commented:
This is the error I get in my logs when trying to build `x64-windows-static`. The error occurs in the debug build.

```
LINK Pass 1: command "C:\PROGRA~2\MICROS~2\2019\PROFES~1\VC\Tools\MSVC\1428~1.299\bin\Hostx64\x64\link.exe src\parquet\CMakeFiles\parquet_shared.dir\arrow\path_internal.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\arrow\reader.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\arrow\reader_internal.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\arrow\schema.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\arrow\schema_internal.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\arrow\writer.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\bloom_filter.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\column_reader.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\column_scanner.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\column_writer.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\encoding.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\encryption\encryption.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\encryption\internal_file_decryptor.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\encryption\internal_file_encryptor.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\exception.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\file_reader.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\file_writer.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\level_comparison.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\level_conversion.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\metadata.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\murmur3.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\__\generated\parquet_constants.cpp.obj src\parquet\CMakeFiles\parquet_shared.dir\__\generated\parquet_types.cpp.obj src\parquet\CMakeFiles\parquet_shared.dir\platform.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\printer.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\properties.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\schema.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\statistics.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\stream_reader.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\stream_writer.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\types.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\level_comparison_avx2.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\level_conversion_bmi2.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\encryption\encryption_internal.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\encryption\crypto_factory.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\encryption\file_key_unwrapper.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\encryption\file_key_wrapper.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\encryption\kms_client.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\encryption\key_material.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\encryption\key_metadata.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\encryption\key_toolkit.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\encryption\key_toolkit_internal.cc.obj src\parquet\CMakeFiles\parquet_shared.dir\encryption\local_wrap_kms_client.cc.obj /out:debug\parquet.dll /implib:debug\parquet.lib /pdb:debug\parquet.pdb /dll /version:400.0 /machine:x64 /NODEFAULTLIB:LIBCMT /nologo /debug /INCREMENTAL debug\arrow.lib D:\Development\Work\vcpkg.ianmcook\installed\x64-windows\lib\thriftmd.lib D:\Development\Work\vcpkg.ianmcook\installed\x64-windows\debug\lib\libssl.lib D:\Development\Work\vcpkg.ianmcook\installed\x64-windows\debug\lib\libcrypto.lib D:\Development\Work\vcpkg.ianmcook\installed\x64-windows\debug\lib\brotlienc.lib D:\Development\Work\vcpkg.ianmcook\installed\x64-windows\debug\lib\brotlidec.lib D:\Development\Work\vcpkg.ianmcook\installed\x64-windows\debug\lib\brotlicommon.lib D:\Development\Work\vcpkg.ianmcook\installed\x64-windows\debug\lib\utf8proc.lib D:\Development\Work\vcpkg.ianmcook\installed\x64-windows\debug\lib\re2.lib mimalloc_ep\src\mimalloc_ep\lib\mimalloc-1.6\mimalloc-static-debug.lib ws2_32.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:src\parquet\CMakeFiles\parquet_shared.dir/intermediate.manifest src\parquet\CMakeFiles\parquet_shared.dir/manifest.res" failed (exit code 1319) with the following output:

thriftmd.lib(TTransportException.cpp.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in path_internal.cc.obj
thriftmd.lib(TTransportException.cpp.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in path_internal.cc.obj
thriftmd.lib(TProtocol.cpp.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in path_internal.cc.obj
thriftmd.lib(TProtocol.cpp.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in path_internal.cc.obj
thriftmd.lib(TBufferTransports.cpp.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in path_internal.cc.obj
thriftmd.lib(TBufferTransports.cpp.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in path_internal.cc.obj
   Creating library debug\parquet.lib and object debug\parquet.exp
LINK : warning LNK4098: defaultlib 'MSVCRT' conflicts with use of other libs; use /NODEFAULTLIB:library
debug\parquet.dll : fatal error LNK1319: 6 mismatches detected
```

I believe that the problem is that it's trying to link Arrow and Parquet with Thrift (Release) `thriftmd.lib` instead of the Debug version `thriftmdd.lib`. MSVC linker correctly detects this. You can see the command line points to the release lib as well: `vcpkg.ianmcook\installed\x64-windows\lib\thriftmd.lib` instead of `vcpkg.ianmcook\installed\x64-windows\debug\lib\thriftmd.lib`

ianmcook commented:
Thanks @GPSnoopy, please feel free to push more commits if you have time to try adding patches. I'm busy with other work today but I'll return to this tomorrow.

JonLiu1993 commented:
@ianmcook ,Thanks for the pr!
The arrow failed on CI testing. could you take a look?  The affected triplets: x86-windows,x64-windows,arm64-windows.

x86-windows
````
D:\buildtrees\arrow\src\rrow-4.0.0-dfe1b4d2b4.clean\cpp\src\arrow/util/bit_util.h(67): error C3861: '__popcnt64': identifier not found
D:\buildtrees\arrow\src\rrow-4.0.0-dfe1b4d2b4.clean\cpp\src\arrow/util/bit_util.h(201): error C3861: '_BitScanReverse64': identifier not found
D:\buildtrees\arrow\src\rrow-4.0.0-dfe1b4d2b4.clean\cpp\src\arrow/util/bit_util.h(247): error C3861: '_BitScanForward64': identifier not found

````

````
CMake Error in CMakeLists.txt:
  You have set CMAKE_BACKWARDS_COMPATIBILITY to a CMake version less than
  2.4.  This version of CMake only supports backwards compatibility with
  CMake 2.4 or later.  For compatibility with older versions please use any
  CMake 2.8.x release or lower.

````

x64-windows
````
hriftmd.lib(TTransportException.cpp.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in path_internal.cc.obj
thriftmd.lib(TTransportException.cpp.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in path_internal.cc.obj
thriftmd.lib(TProtocol.cpp.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in path_internal.cc.obj
thriftmd.lib(TProtocol.cpp.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in path_internal.cc.obj
thriftmd.lib(TBufferTransports.cpp.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in path_internal.cc.obj
thriftmd.lib(TBufferTransports.cpp.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in path_internal.cc.obj
   Creating library debug\parquet.lib and object debug\parquet.exp

````

arm64-windows
````
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29910\include\nmmintrin.h(17): fatal error C1189: #error:  This header is specific to X86 and X64 targets
[21/253] C:\PROGRA~2\MICROS~1\2019\ENTERP~1\VC\Tools\MSVC\1428~1.299\bin\Hostx64\arm64\cl.exe   /TP -DARROW_EXPORTING -DARROW_HAVE_RUNTIME_SSE4_2 -DARROW_HAVE_SSE4_2 -DARROW_MIMALLOC -DARROW_WITH_BROTLI -DARROW_WITH_BZ2 -DARROW_WITH_LZ4 -DARROW_WITH_RE2 -DARROW_WITH_SNAPPY -DARROW_WITH_TIMING_TESTS -DARROW_WITH_UTF8PROC -DARROW_WITH_ZLIB -DARROW_WITH_ZSTD -DBOOST_ALL_DYN_LINK -DBOOST_ALL_NO_LIB -DURI_STATIC_BUILD -D_CRT_SECURE_NO_WARNINGS -Darrow_shared_EXPORTS -Isrc -ID:\buildtrees\arrow\src\rrow-4.0.0-dfe1b4d2b4.clean\cpp\src -ID:\buildtrees\arrow\src\rrow-4.0.0-dfe1b4d2b4.clean\cpp\src\generated -ID:\buildtrees\arrow\src\rrow-4.0.0-dfe1b4d2b4.clean\cpp\thirdparty\flatbuffers\include -ID:\installed\arm64-windows\include -Imimalloc_ep\src\mimalloc_ep\lib\mimalloc-1.6\include -ID:\installed\arm64-windows\share\rapidjson\..\..\include -Ixsimd_ep\src\xsimd_ep-install\include -ID:\buildtrees\arrow\src\rrow-4.0.0-dfe1b4d2b4.clean\cpp\thirdparty\hadoop\include /nologo /DWIN32 /D_WINDOWS  /utf-8 /GR /EHsc /MP  /D_SILENCE_TR1_NAMESPACE_DEPRECATION_WARNING   /EHsc /wd5105 /bigobj /utf-8 /W3 /wd4800 /wd4996 /wd4065   /D_DEBUG /MDd /Z7 /Ob0 /Od /RTC1 /showIncludes /Fosrc\arrow\CMakeFiles\arrow_shared.dir\compute\kernels\codegen_internal.cc.obj /Fdsrc\arrow\CMakeFiles\arrow_shared.dir\ /FS -c D:\buildtrees\arrow\src\rrow-4.0.0-dfe1b4d2b4.clean\cpp\src\arrow\compute\kernels\codegen_internal.cc
FAILED: src/arrow/CMakeFiles/arrow_shared.dir/compute/kernels/codegen_internal.cc.obj 
C:\PROGRA~2\MICROS~1\2019\ENTERP~1\VC\Tools\MSVC\1428~1.299\bin\Hostx64\arm64\cl.exe   /TP -DARROW_EXPORTING -DARROW_HAVE_RUNTIME_SSE4_2 -DARROW_HAVE_SSE4_2 -DARROW_MIMALLOC -DARROW_WITH_BROTLI -DARROW_WITH_BZ2 -DARROW_WITH_LZ4 -DARROW_WITH_RE2 -DARROW_WITH_SNAPPY -DARROW_WITH_TIMING_TESTS -DARROW_WITH_UTF8PROC -DARROW_WITH_ZLIB -DARROW_WITH_ZSTD -DBOOST_ALL_DYN_LINK -DBOOST_ALL_NO_LIB -DURI_STATIC_BUILD -D_CRT_SECURE_NO_WARNINGS -Darrow_shared_EXPORTS -Isrc -ID:\buildtrees\arrow\src\rrow-4.0.0-dfe1b4d2b4.clean\cpp\src -ID:\buildtrees\arrow\src\rrow-4.0.0-dfe1b4d2b4.clean\cpp\src\generated -ID:\buildtrees\arrow\src\rrow-4.0.0-dfe1b4d2b4.clean\cpp\thirdparty\flatbuffers\include -ID:\installed\arm64-windows\include -Imimalloc_ep\src\mimalloc_ep\lib\mimalloc-1.6\include -ID:\installed\arm64-windows\share\rapidjson\..\..\include -Ixsimd_ep\src\xsimd_ep-install\include -ID:\buildtrees\arrow\src\rrow-4.0.0-dfe1b4d2b4.clean\cpp\thirdparty\hadoop\include /nologo /DWIN32 /D_WINDOWS  /utf-8 /GR /EHsc /MP  /D_SILENCE_TR1_NAMESPACE_DEPRECATION_WARNING   /EHsc /wd5105 /bigobj /utf-8 /W3 /wd4800 /wd4996 /wd4065   /D_DEBUG /MDd /Z7 /Ob0 /Od /RTC1 /showIncludes /Fosrc\arrow\CMakeFiles\arrow_shared.dir\compute\kernels\codegen_internal.cc.obj /Fdsrc\arrow\CMakeFiles\arrow_shared.dir\ /FS -c D:\buildtrees\arrow\src\rrow-4.0.0-dfe1b4d2b4.clean\cpp\src\arrow\compute\kernels\codegen_internal.cc
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.28.29910\include\nmmintrin.h(17): fatal error C1189: #error:  This header is specific to X86 and X64 targets
[22/253] C:\PROGRA~2\MICROS~1\2019\ENTERP~1\VC\Tools\MSVC\1428~1.299\bin\Hostx64\arm64\cl.exe   /TP -DARROW_EXPORTING -DARROW_HAVE_RUNTIME_SSE4_2 -DARROW_HAVE_SSE4_2 -DARROW_MIMALLOC -DARROW_WITH_BROTLI -DARROW_WITH_BZ2 -DARROW_WITH_LZ4 -DARROW_WITH_RE2 -DARROW_WITH_SNAPPY -DARROW_WITH_TIMING_TESTS -DARROW_WITH_UTF8PROC -DARROW_WITH_ZLIB -DARROW_WITH_ZSTD -DBOOST_ALL_DYN_LINK -DBOOST_ALL_NO_LIB -DURI_STATIC_BUILD -D_CRT_SECURE_NO_WARNINGS -Darrow_shared_EXPORTS -Isrc -ID:\buildtrees\arrow\src\rrow-4.0.0-dfe1b4d2b4.clean\cpp\src -ID:\buildtrees\arrow\src\rrow-4.0.0-dfe1b4d2b4.clean\cpp\src\generated -ID:\buildtrees\arrow\src\rrow-4.0.0-dfe1b4d2b4.clean\cpp\thirdparty\flatbuffers\include -ID:\installed\arm64-windows\include -Imimalloc_ep\src\mimalloc_ep\lib\mimalloc-1.6\include -ID:\installed\arm64-windows\share\rapidjson\..\..\include -Ixsimd_ep\src\xsimd_ep-install\include -ID:\buildtrees\arrow\src\rrow-4.0.0-dfe1b4d2b4.clean\cpp\thirdparty\hadoop\include /nologo /DWIN32 /D_WINDOWS  /utf-8 /GR /EHsc /MP  /D_SILENCE_TR1_NAMESPACE_DEPRECATION_WARNING   /EHsc /wd5105 /bigobj /utf-8 /W3 /wd4800 /wd4996 /wd4065   /D_DEBUG /MDd /Z7 /Ob0 /Od /RTC1 /showIncludes /Fosrc\arrow\CMakeFiles\arrow_shared.dir\compute\kernels\scalar_cast_boolean.cc.obj /Fdsrc\arrow\CMakeFiles\arrow_shared.dir\ /FS -c D:\buildtrees\arrow\src\rrow-4.0.0-dfe1b4d2b4.clean\cpp\src\arrow\compute\kernels\scalar_cast_boolean.cc
FAILED: src/arrow/CMakeFiles/arrow_shared.dir/compute/kernels/scalar_cast_boolean.cc.obj 
````

ianmcook committed:
Apply patches

ianmcook committed:
Remove LIB_DIR_OPTIONS

ianmcook committed:
Merge branch 'master' of github.com:ianmcook/vcpkg

ianmcook committed:
Tweak zstd flags

ianmcook committed:
Update version hash

ianmcook committed:
Format

ianmcook committed:
Fail early on x86

ianmcook committed:
Update hash

ianmcook committed:
Fail early on arm, arm64

ianmcook committed:
Update hash

ianmcook committed:
Add expected failures to to scripts/ci.baseline.txt

ianmcook commented:
I updated the 3.0.0 patches to work with 4.0.0, removed some CMake options that were obviated by the patches, debugged a zstd issue, and specified expected failures on x86, arm, and arm64.

@GPSnoopy please retest on your end, thanks!

GPSnoopy commented:
Hi @ianmcook,

I've started a ParquetSharp PR pointing at your vcpkg port (this uses `x64-windows-static`, `x64-linux` and `x64-osx`).
See: https://github.com/G-Research/ParquetSharp/pull/198

After fixing a few breaking changes, it seems that the compilation is stuck on `Mimalloc` linker issue. I suspect it's because Arrow internally ships its own version of the allocator library, yet the `.lib` file is not made available for a client application to use.

One obvious fix is to use the existing `mimalloc` vcpkg port. If the arrow port were to publish its own mimalloc static library, this could cause problems with applications that already links with vcpkg mimalloc.

JonLiu1993 commented:
@ianmcook ，If this pr is ready for review, please switch to open state. Thank you

ianmcook committed:
Exclude mallocs from default features

ianmcook committed:
Update hash

ianmcook commented:
@GPSnoopy I'll look into that, thanks. I believe Arrow needs to set some of the CMake flags to non-default values when building mimalloc and jemalloc so I'm not sure what would be involved in trying to use the mimalloc and jemalloc vcpkg ports instead of the vendored versions in Arrow. I believe the vendored versions in Arrow have been modified to avoid conflicts but I'm not sure of the details of that.

As an immediate fix, I simply removed mimalloc and jemalloc from the default features so unless you specify them they will not build. Does that solve the ParquetSharp linker issue you observed?

GPSnoopy commented:
@ianmcook ParquetSharp PR is now green and all unit tests pass. 😃

ianmcook commented:
@JonLiu1993 ready for review now, thanks!

ras0219-msft commented on ports/arrow/portfile.cmake:
```diff
@@ -3,53 +3,72 @@ vcpkg_fail_port_install(ON_ARCH "x86" "arm" "arm64")
+        -DARROW_BUILD_STATIC=${ARROW_BUILD_STATIC}
+        -DARROW_BUILD_TESTS=OFF
+        -DARROW_DEPENDENCY_SOURCE=SYSTEM
+        -DARROW_DEPENDENCY_USE_SHARED=${ARROW_DEPENDENCY_USE_SHARED}
```
```suggestion
        -DARROW_DEPENDENCY_USE_SHARED=OFF
```

Arrow cannot assume that its dependencies are built with any particular linkage -- instead, we should pass a constant here (ON or OFF) and patch the buildsystem to be agnostic.

ras0219-msft commented on ports/arrow/portfile.cmake:
```diff
@@ -3,53 +3,72 @@ vcpkg_fail_port_install(ON_ARCH "x86" "arm" "arm64")
+
+string(COMPARE EQUAL ${VCPKG_LIBRARY_LINKAGE} "dynamic" ARROW_BUILD_SHARED)
+string(COMPARE EQUAL ${VCPKG_LIBRARY_LINKAGE} "static" ARROW_BUILD_STATIC)
+string(COMPARE EQUAL ${VCPKG_LIBRARY_LINKAGE} "dynamic" ARROW_DEPENDENCY_USE_SHARED)
```
```suggestion
```

See below

ras0219-msft commented on ports/arrow/portfile.cmake:
```diff
@@ -3,53 +3,72 @@ vcpkg_fail_port_install(ON_ARCH "x86" "arm" "arm64")
+        -DARROW_WITH_ZSTD=ON
+        -DZSTD_INCLUDE_DIR=${CURRENT_INSTALLED_DIR}/include
+        -DZSTD_MSVC_LIB_PREFIX=
+        -DZSTD_ROOT=${CURRENT_INSTALLED_DIR}
```
This is likely to link the release version of the dependency while in debug mode; could you please check that the correct debug dependency is being linked?

ianmcook replied:
Ok, thanks, I'll check it. To clarify: by "this" are you referring only to the line that sets `ZSTD_ROOT`?

ianmcook replied:
After running some tests, I think it's fine to remove `ZSTD_INCLUDE_DIR` and `DZSTD_ROOT`, so I did that.

`ZSTD_MSVC_LIB_PREFIX` is required. Looking at the relevant Arrow CMake script (https://github.com/apache/arrow/blob/master/cpp/cmake_modules/Findzstd.cmake), I don't believe this interferes with debug/release linking.

ras0219-msft commented on ports/arrow/vcpkg.json:
```diff
@@ -0,0 +1,87 @@
+  ],
+  "default-features": [
+    "csv",
+    "dataset",
```
```suggestion
```
See https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#default-features-should-enable-behaviors-not-apis -- we want to minimize default features as much as possible; they should not be used just to enable things that users might likely want to use.

ras0219-msft commented on ports/arrow/vcpkg.json:
```diff
@@ -0,0 +1,87 @@
+      "description": "S3 support",
+      "dependencies": [
+        {
+          "name": "aws-sdk-cpp",
```
```suggestion
          "name": "aws-sdk-cpp",
          "default-features": false,
```
Disabling the defaults is needed to avoid an unexpected dependency on dynamo.

ras0219-msft commented on ports/arrow/vcpkg.json:
```diff
@@ -0,0 +1,87 @@
+  "name": "arrow",
+  "version": "4.0.0",
+  "description": "Cross-language development platform for in-memory analytics",
+  "homepage": "https://arrow.apache.org",
```
```suggestion
  "homepage": "https://arrow.apache.org",
  "supports": "x64",
```
The previous supports clause should be preserved

ras0219-msft commented on scripts/ci.baseline.txt:
```diff
@@ -47,6 +47,9 @@
 # other ports
 # Cross compiling CI machine cannot run gen_test_char to generate apr_escape_test_char.h
 apr:arm64-windows=fail
+arrow:arm64-windows=fail
+arrow:arm-uwp=fail
+arrow:x86-windows=fail
```
```suggestion
```

Adding back the `"supports"` field removes the need for these explicit failures.

ras0219-msft reviewed (changes requested):
Thanks for the update PR! Please take a look at my comments.

ianmcook committed:
Set default-features to false for aws-sdk-cpp

Co-authored-by: Robert Schumacher <roschuma@microsoft.com>

ianmcook committed:
Specify only x64 support in manifest

Co-authored-by: Robert Schumacher <roschuma@microsoft.com>

ianmcook committed:
Remove unneeded ci.baseline.txt entries

Co-authored-by: Robert Schumacher <roschuma@microsoft.com>

ianmcook committed:
Remove dataset from default-features

Co-authored-by: Robert Schumacher <roschuma@microsoft.com>

ianmcook committed:
Update hash

ianmcook committed:
Remove zstd path args

ianmcook committed:
Update hash

ianmcook commented:
@GPSnoopy could you please confirm that the changes I've made today have not introduced any new issues for ParquetSharp? Thanks

GPSnoopy commented:
Still green. All good! 👍

ianmcook commented:
@ras0219-msft please let me know if the changes made and replies to your comments are satisfactory. Thank you 🙏

draveness commented:
is it easier to use arrow as a static dependency after we merge this? 🤣

draveness commented:
and btw, I think arrow got a 4.0.1 version. Can we update to that directly, or it is a major change

GPSnoopy commented:
@draveness See the related ParquetSharp PR. That library uses arrow as a static library dependency

draveness commented:
> @draveness See the related ParquetSharp PR. That library uses arrow as a static library dependency

Oh, this one? https://github.com/G-Research/ParquetSharp/pull/198/files

ianmcook commented:
@draveness I think it would be best to update to 4.0.1 in a separate PR. That way there is a version 4.0.0 in the vcpkg version history.

@ras0219-msft @JonLiu1993 please let me know what if anything I can do to help get this merged. Thank you

ras0219-msft commented:
This LGTM, thanks for all the work here :)
